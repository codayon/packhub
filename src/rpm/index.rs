use askama::Template;

use super::package::{Dependency, RPMPackage};

#[derive(Debug, Template)]
#[template(path = "primary.xml")]
struct Primary {
    packages: Vec<RPMPackage>,
}

#[derive(Debug, Template)]
#[template(path = "filelists.xml")]
struct FileLists {
    packages: Vec<RPMPackage>,
}

#[derive(Debug, Template)]
#[template(path = "other.xml")]
struct Other {
    packages: Vec<RPMPackage>,
}

pub fn get_primary_index(packages: Vec<RPMPackage>) -> String {
    let primary = Primary { packages };
    primary.render().unwrap()
}

pub fn get_filelists_index(packages: Vec<RPMPackage>) -> String {
    let list = FileLists { packages };
    list.render().unwrap()
}

pub fn get_other_index(packages: Vec<RPMPackage>) -> String {
    let list = Other { packages };
    list.render().unwrap()
}

#[cfg(test)]
mod tests {
    use std::fs::read;

    use chrono::DateTime;
    use insta::assert_snapshot;

    use crate::package::Package;

    use super::*;

    #[test]
    fn test_rpm_indices() {
        let package = Package::detect_package("OpenBangla-Keyboard_2.0.0-fedora38.rpm", "2.0.0".to_owned(), "https://github.com/OpenBangla/OpenBangla-Keyboard/releases/download/2.0.0/OpenBangla-Keyboard_2.0.0-fedora38.rpm".to_owned(), DateTime::UNIX_EPOCH).unwrap();
        let data = read("data/OpenBangla-Keyboard_2.0.0-fedora38.rpm").unwrap();
        package.set_data(data);
        let package = RPMPackage::from_package(&package).unwrap();
        let packages = vec![package];

        assert_snapshot!(get_primary_index(packages.clone()));

        assert_snapshot!(get_filelists_index(packages.clone()));

        assert_snapshot!(get_other_index(packages));
    }
}
