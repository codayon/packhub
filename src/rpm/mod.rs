mod index;
mod package;
mod routes;

pub use self::routes::rpm_routes;
