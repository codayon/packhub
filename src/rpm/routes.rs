use axum::{
    body::Body, extract::Path, http::StatusCode, response::IntoResponse, routing::get, Router,
};
use axum_extra::{headers::UserAgent, typed_header::TypedHeader};
use zstd::encode_all;

use crate::{
    pgp::{detached_sign_metadata, load_secret_key_from_file},
    repository::Repository,
    rpm::{index::get_repomd_index, package::RPMPackage},
    utils::download_packages,
};

use super::index::{get_filelists_index, get_other_index, get_primary_index};

async fn index(
    Path((owner, repo, file)): Path<(String, String, String)>,
    TypedHeader(agent): TypedHeader<UserAgent>,
) -> Result<Vec<u8>, StatusCode> {
    let repo = Repository::from_github(owner, repo).await;
    let Some(packages) = repo.select_package_rpm(agent.as_str()) else {
        return Err(StatusCode::NOT_FOUND);
    };

    let packages = packages.into_iter().map(|p| p.clone()).collect();

    let packages: Vec<RPMPackage> = download_packages(packages)
        .await
        .unwrap()
        .into_iter()
        .map(|p| RPMPackage::from_package(&p).unwrap())
        .collect();

    match file.as_str() {
        "repomd.xml" => Ok(get_repomd_index(&packages).into_bytes()),
        "repomd.xml.asc" => {
            let metadata = get_repomd_index(&packages);
            let secret_key = load_secret_key_from_file().unwrap();
            let signature = detached_sign_metadata("repomd.xml", &metadata, &secret_key)
                .unwrap()
                .into_bytes();
            Ok(signature)
        }
        "primary.xml.zst" => Ok(encode_all(get_primary_index(&packages).as_bytes(), 0).unwrap()),
        "filelists.xml.zst" => {
            Ok(encode_all(get_filelists_index(&packages).as_bytes(), 0).unwrap())
        }
        "other.xml.zst" => Ok(encode_all(get_other_index(&packages).as_bytes(), 0).unwrap()),
        _ => Err(StatusCode::NOT_FOUND),
    }
}

async fn package(
    Path((owner, repo, ver, file)): Path<(String, String, String, String)>,
) -> impl IntoResponse {
    let url = format!("https://github.com/{owner}/{repo}/releases/download/{ver}/{file}");
    let res = reqwest::get(url).await.unwrap();
    let stream = res.bytes_stream();
    let stream = Body::from_stream(stream);

    stream
}

pub fn rpm_routes() -> Router {
    Router::new()
        .route("/github/:owner/:repo/repodata/:file", get(index))
        .route("/github/:owner/:repo/package/:ver/:file", get(package))
}
