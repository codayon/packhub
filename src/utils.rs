use std::ops::Add;

use sha1::digest::{generic_array::ArrayLength, Digest, OutputSizeUser};

pub fn hashsum<T: Digest>(data: &[u8]) -> String
where
    <T as OutputSizeUser>::OutputSize: Add,
    <<T as OutputSizeUser>::OutputSize as Add>::Output: ArrayLength<u8>,
{
    format!("{:x}", T::digest(data))
}
