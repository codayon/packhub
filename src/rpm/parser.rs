use anyhow::{Context, Result};
use rpm::{DependencyFlags, FileMode, IndexSignatureTag};
use sha2::Sha256;

use crate::utils::hashsum;

#[derive(Debug)]
struct RPM {
    name: String,
    epoch: u32,
    version: String,
    release: String,
    arch: String,
    vendor: String,
    url: String,
    license: String,
    summary: String,
    description: String,
    group: String,
    build_time: u64,
    build_host: String,
    source: String,
    provides: Vec<Dependency>,
    requires: Vec<Dependency>,
    sha256: String,
    header_start: u64,
    header_end: u64,
    files: Vec<File>,
    pkg_size: usize,
    installed_size: u64,
    archive_size: u64,
}

#[derive(Debug)]
struct Dependency {
    dependency: String,
    version: String,
    condition: String,
}

#[derive(Debug)]
struct File {
    path: String,
    dir: bool,
}

pub fn parse_rpm(mut data: &[u8]) -> Result<RPM> {
    let pkg_size = data.len();
    let rpm = rpm::Package::parse(&mut data)
        .context("Unable to parse the package using rpm parser crate")?;
    let header = rpm.metadata;

    let name = header.get_name()?.to_owned();
    let epoch = header.get_epoch().unwrap_or(0);
    let version = header.get_version()?.to_owned();
    let release = header.get_release()?.to_owned();
    let arch = header.get_arch()?.to_owned();
    let vendor = header.get_vendor()?.to_owned();
    let url = header.get_url()?.to_owned();
    let license = header.get_license()?.to_owned();
    let summary = header.get_summary()?.to_owned();
    let description = header.get_description()?.to_owned();
    let group = header.get_group()?.to_owned();
    let build_time = header.get_build_time()?;
    let build_host = header.get_build_host()?.to_owned();
    let source = header.get_source_rpm()?.to_owned();
    let sha256 = hashsum::<Sha256>(data);
    let range = header.get_package_segment_offsets();
    let header_start = range.header;
    let header_end = range.payload;
    let installed_size = header.get_installed_size()?;
    let archive_size = header
        .signature
        .get_entry_data_as_u64(IndexSignatureTag::RPMSIGTAG_LONGARCHIVESIZE)
        .or_else(|_e| {
            header
                .signature
                .get_entry_data_as_u32(IndexSignatureTag::RPMSIGTAG_PAYLOADSIZE)
                .map(|v| v as u64)
        })?;

    let provides: Vec<Dependency> = header
        .get_provides()?
        .into_iter()
        .map(|i| Dependency {
            dependency: i.name,
            version: i.version,
            condition: flag_to_condition(i.flags),
        })
        .collect();

    let requires: Vec<Dependency> = header
        .get_requires()?
        .into_iter()
        .filter(|i| !i.flags.contains(DependencyFlags::RPMLIB))
        .filter(|i| !i.flags.contains(DependencyFlags::CONFIG))
        .map(|i| Dependency {
            dependency: i.name,
            version: i.version,
            condition: flag_to_condition(i.flags),
        })
        .collect();
    
    let files: Vec<File> = header
        .get_file_entries()?
        .into_iter()
        .map(|i| File {
            path: i.path.to_string_lossy().to_string(),
            dir: matches!(i.mode, FileMode::Dir { .. }),
        })
        .collect();

    Ok(RPM {
        name,
        epoch,
        version,
        release,
        arch,
        vendor,
        url,
        license,
        summary,
        description,
        group,
        build_time,
        build_host,
        source,
        provides,
        requires,
        sha256,
        header_start,
        header_end,
        files,
        pkg_size,
        installed_size,
        archive_size,
    })
}

fn flag_to_condition(flags: DependencyFlags) -> String {
    if flags.contains(DependencyFlags::GE) {
        "GE".to_owned()
    } else if flags.contains(DependencyFlags::LE) {
        "LE".to_owned()
    } else if flags.contains(DependencyFlags::EQUAL) {
        "EQ".to_owned()
    } else {
        "".to_owned()
    }
}

#[cfg(test)]
mod tests {
    use std::fs::read;

    use insta::assert_debug_snapshot;

    use super::*;

    #[test]
    fn test_parser() {
        let data = read("data/OpenBangla-Keyboard_2.0.0-fedora38.rpm").unwrap();
        let parsed = parse_rpm(&data[..]).unwrap();
        assert_debug_snapshot!(parsed);
    }
}
