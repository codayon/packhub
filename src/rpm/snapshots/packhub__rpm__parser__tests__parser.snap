---
source: src/rpm/parser.rs
expression: parsed
---
RPM {
    name: "openbangla-keyboard",
    epoch: 0,
    version: "2.0.0",
    release: "1.fc38",
    arch: "x86_64",
    vendor: "OpenBangla Team",
    url: "https://openbangla.github.io/",
    license: "GPLv3",
    summary: "OpenSource Bengali input method",
    description: "OpenBangla Keyboard is an OpenSource, Unicode compliant Bengali Input Method for GNU/Linux systems. It's a full fledged Bengali input method with typing automation tools, includes many famous typing methods such as Avro Phonetic, Probhat, Munir Optima, National (Jatiya) etc.\n\nMost of the features of Avro Keyboard are present in OpenBangla Keyboard. So Avro Keyboard users will feel at home with OpenBangla Keyboard in Linux.",
    group: "unknown",
    build_time: 1684606381,
    build_host: "3aebfc71f602",
    source: "openbangla-keyboard-2.0.0-1.fc38.src.rpm",
    provides: [
        Dependency {
            dependency: "application()",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "application(openbangla-keyboard.desktop)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "config(openbangla-keyboard)",
            version: "2.0.0-1.fc38",
            condition: "EQ",
        },
        Dependency {
            dependency: "metainfo()",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "metainfo(io.github.openbangla.keyboard.metainfo.xml)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "openbangla-keyboard",
            version: "2.0.0-1.fc38",
            condition: "EQ",
        },
        Dependency {
            dependency: "openbangla-keyboard(x86-64)",
            version: "2.0.0-1.fc38",
            condition: "EQ",
        },
    ],
    requires: [
        Dependency {
            dependency: "ibus",
            version: "1.5.1",
            condition: "GE",
        },
        Dependency {
            dependency: "ibus-libs",
            version: "1.5.1",
            condition: "GE",
        },
        Dependency {
            dependency: "ld-linux-x86-64.so.2()(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "ld-linux-x86-64.so.2(GLIBC_2.3)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libQt5Core.so.5()(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libQt5Core.so.5(Qt_5)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libQt5Core.so.5(Qt_5.15)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libQt5Gui.so.5()(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libQt5Gui.so.5(Qt_5)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libQt5Network.so.5()(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libQt5Network.so.5(Qt_5)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libQt5Widgets.so.5()(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libQt5Widgets.so.5(Qt_5)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6()(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.10)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.14)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.15)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.17)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.18)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.2.5)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.25)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.27)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.28)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.29)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.3)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.3.4)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.32)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.33)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.34)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.4)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.5)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.6)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libc.so.6(GLIBC_2.9)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libgcc_s.so.1()(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libgcc_s.so.1(GCC_3.0)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libgcc_s.so.1(GCC_3.3)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libgcc_s.so.1(GCC_4.2.0)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libgio-2.0.so.0()(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libglib-2.0.so.0()(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libgobject-2.0.so.0()(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libibus-1.0.so.5()(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libm.so.6()(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libm.so.6(GLIBC_2.2.5)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libstdc++.so.6()(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libstdc++.so.6(CXXABI_1.3)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libstdc++.so.6(CXXABI_1.3.1)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libstdc++.so.6(GLIBCXX_3.4)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libstdc++.so.6(GLIBCXX_3.4.21)(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "libzstd",
            version: "1.3.3",
            condition: "GE",
        },
        Dependency {
            dependency: "libzstd.so.1()(64bit)",
            version: "",
            condition: "",
        },
        Dependency {
            dependency: "qt5-qtbase",
            version: "5.9.0",
            condition: "GE",
        },
        Dependency {
            dependency: "rtld(GNU_HASH)",
            version: "",
            condition: "",
        },
    ],
    sha256: "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    header_start: 4504,
    header_end: 14821,
    files: [
        File {
            path: "/usr/lib/.build-id",
            dir: true,
        },
        File {
            path: "/usr/lib/.build-id/45",
            dir: true,
        },
        File {
            path: "/usr/lib/.build-id/45/4f3de1c02f67096ca79f7edb9030f7b1b4d015",
            dir: false,
        },
        File {
            path: "/usr/lib/.build-id/b2",
            dir: true,
        },
        File {
            path: "/usr/lib/.build-id/b2/1aa073a506f8b918fda931c8d12912b5d164be",
            dir: false,
        },
        File {
            path: "/usr/share/applications/openbangla-keyboard.desktop",
            dir: false,
        },
        File {
            path: "/usr/share/ibus/component/openbangla.xml",
            dir: false,
        },
        File {
            path: "/usr/share/icons/hicolor/1024x1024/apps/openbangla-keyboard.png",
            dir: false,
        },
        File {
            path: "/usr/share/icons/hicolor/128x128/apps/openbangla-keyboard.png",
            dir: false,
        },
        File {
            path: "/usr/share/icons/hicolor/16x16/apps/openbangla-keyboard.png",
            dir: false,
        },
        File {
            path: "/usr/share/icons/hicolor/32x32/apps/openbangla-keyboard.png",
            dir: false,
        },
        File {
            path: "/usr/share/icons/hicolor/48x48/apps/openbangla-keyboard.png",
            dir: false,
        },
        File {
            path: "/usr/share/icons/hicolor/512x512/apps/openbangla-keyboard.png",
            dir: false,
        },
        File {
            path: "/usr/share/metainfo/io.github.openbangla.keyboard.metainfo.xml",
            dir: false,
        },
        File {
            path: "/usr/share/openbangla-keyboard",
            dir: true,
        },
        File {
            path: "/usr/share/openbangla-keyboard/data",
            dir: true,
        },
        File {
            path: "/usr/share/openbangla-keyboard/data/autocorrect.json",
            dir: false,
        },
        File {
            path: "/usr/share/openbangla-keyboard/data/dictionary.json",
            dir: false,
        },
        File {
            path: "/usr/share/openbangla-keyboard/data/regex.json",
            dir: false,
        },
        File {
            path: "/usr/share/openbangla-keyboard/data/suffix.json",
            dir: false,
        },
        File {
            path: "/usr/share/openbangla-keyboard/ibus-openbangla",
            dir: false,
        },
        File {
            path: "/usr/share/openbangla-keyboard/icons",
            dir: true,
        },
        File {
            path: "/usr/share/openbangla-keyboard/icons/OpenBangla-Keyboard.png",
            dir: false,
        },
        File {
            path: "/usr/share/openbangla-keyboard/layouts",
            dir: true,
        },
        File {
            path: "/usr/share/openbangla-keyboard/layouts/Avro_Easy.json",
            dir: false,
        },
        File {
            path: "/usr/share/openbangla-keyboard/layouts/Borno.json",
            dir: false,
        },
        File {
            path: "/usr/share/openbangla-keyboard/layouts/Munir_Optima.json",
            dir: false,
        },
        File {
            path: "/usr/share/openbangla-keyboard/layouts/National_Jatiya.json",
            dir: false,
        },
        File {
            path: "/usr/share/openbangla-keyboard/layouts/Probhat.json",
            dir: false,
        },
        File {
            path: "/usr/share/openbangla-keyboard/layouts/avrophonetic.json",
            dir: false,
        },
        File {
            path: "/usr/share/openbangla-keyboard/openbangla-gui",
            dir: false,
        },
        File {
            path: "/usr/share/pixmaps/openbangla-keyboard.png",
            dir: false,
        },
    ],
    pkg_size: 4615096,
    installed_size: 12106690,
    archive_size: 12112024,
}
